# TMLE — causal effect of a modifiable dropout driver

**Design Spec §5.3.** Third and final stage of the analytics progression. Unlike Cox PH and RSF
(both predictive), this stage asks a causal question: does a specific, actionable factor *cause*
patients to drop out, adjusting for confounding — the number that would actually justify a
retention intervention.

## Estimand (ICH E9(R1) language, stated before fitting anything per Design Spec §5.1/§9)

**Population:** all pooled patients across the 3 raw-domain trials (PACCE, FOLFIRI, FOLFOX/PRIME)
with non-missing covariates.

**Exposure (the modifiable driver):** `A = 1` if the patient experienced at least one serious
adverse event (`serious == 'Y'` in `stg_ae`) prior to their dropout/end of observed follow-up;
`A = 0` otherwise. Chosen over the Design Spec's example (`site distance`) because no site/
geographic covariate exists in the loaded PDS raw domains — SAE occurrence is the strongest
*actually available* modifiable driver, and it's operationally actionable (proactive monitoring,
dose modification, supportive care), matching the agent layer's framing in Design Spec §6.

**Outcome:** `Y = 1` if the patient's `discontinuation_reason` reflects **non-clinical
discontinuation** — Adverse event, Subject request, Administrative decision, Consent withdrawn,
Noncompliance, Other, Protocol deviation, Lost to follow-up — i.e. reasons a retention
intervention could plausibly have prevented. `Y = 0` if still on study, or discontinued for
Disease progression or Death (the drug's efficacy running out, not the patient leaving the
study — a distinct concept from the survival endpoint used in the Cox/RSF stages, and exactly the
modifiable-vs-non-modifiable distinction Design Spec §6's agent layer is built around).

**Strategy for handling the dropout event itself (the ICH E9(R1) intercurrent-event question):**
a **treatment policy strategy** — the estimand is the effect of the SAE on non-clinical
discontinuation as it actually occurs in this population, not a hypothetical "what would have
happened had the patient never been able to drop out" world. This is the natural choice for a
retention-intervention use case: the whole point is estimating the real-world effect size a
coordinator-facing intervention would be acting against.

**Target parameter:** average treatment effect (risk difference, risk ratio, odds ratio) of `A` on
`Y`, adjusted for the same demographic/clinical confounders used in the Cox/RSF stages
(`age`, `sex`, `race_category`, `ecog_baseline`, `diagnosis_type`, `bsa_m2`, `on_panitumumab`,
`trial_id`), estimated via TMLE (`zepid`) for its doubly-robust property — the estimate stays
consistent if *either* the exposure or outcome nuisance model is correctly specified, not both.

**Scope limitation, stated explicitly rather than glossed over:** `zepid`'s TMLE implementation is
a parametric, single-timepoint estimator (fixed exposure/outcome regression formulas), not a
cross-fitted/machine-learning-nuisance or longitudinal TMLE. Design Spec §5.3's "`trial_id` as a
cross-validation fold/cluster variable" is approximated here by including `trial_id` as a
covariate in both nuisance models, adjusting for cross-trial differences directly rather than via
a genuine cluster-robust variance estimator or cross-fitting procedure — a real simplification for
this sprint's scope, not a claim of full methodological rigor. `kras_status` is excluded from both
nuisance models for the same structural reason established in the Cox PH notebook: its
`"Unknown"` category is a perfect proxy for FOLFOX/PRIME trial membership, which would collide
directly with including `trial_id` as a covariate.

In [2]:
import warnings
warnings.filterwarnings('ignore')  # zepid's TMLE relies on the deprecated pkg_resources API

import numpy as np
import pandas as pd
from sqlalchemy import create_engine

pd.set_option('display.max_columns', None)

engine = create_engine(
    "postgresql+psycopg2://VincenzoVantornout:localdevpassword@localhost:5432/postgres"
)

patients = pd.read_sql("select * from public.fct_patient_survival", engine)
ae = pd.read_sql("select trial_id, subjid, serious, start_day from public.stg_ae", engine)

print(f"{len(patients)} patients, {len(ae)} adverse event rows")

2723 patients, 66723 adverse event rows


## Building the exposure — with correct temporal ordering

A naive "did this patient ever have a serious AE" ignores *when* — if the SAE happened after the
patient already dropped out, it can't have caused the dropout, and including it would be reverse
causation, not a modifiable driver. `discontinuation_day` (added to `stg_disposition` for this
notebook — see `learning-notes/dbt.md`) fixes this: exposure is scoped to SAEs occurring at or
before each patient's relevant reference day — `discontinuation_day` if they discontinued, or
their overall observed `time_days` otherwise (their full observed follow-up window).

In [3]:
model_df = patients.copy()
model_df['ref_day'] = model_df['discontinuation_day'].fillna(model_df['time_days'])

serious_ae = ae[ae['serious'] == 'Y'][['trial_id', 'subjid', 'start_day']]
exposed = serious_ae.merge(
    model_df[['trial_id', 'subjid', 'ref_day']], on=['trial_id', 'subjid'], how='inner'
)
exposed = exposed[exposed['start_day'] <= exposed['ref_day']]
exposed_ids = exposed[['trial_id', 'subjid']].drop_duplicates()
exposed_ids['A'] = 1

model_df = model_df.merge(exposed_ids, on=['trial_id', 'subjid'], how='left')
model_df['A'] = model_df['A'].fillna(0).astype(int)

print(f"Exposed (A=1, serious AE before dropout/end of follow-up): {model_df['A'].sum()}")
print(f"Unexposed (A=0): {(model_df['A'] == 0).sum()}")

Exposed (A=1, serious AE before dropout/end of follow-up): 917
Unexposed (A=0): 1806


## Building the outcome — non-clinical discontinuation

`discontinuation_reason` is blank/empty for patients still on study (per
`dropout_reason_harmonization.ipynb`'s value counts), not `NaN` — handled explicitly below rather
than assuming one or the other.

In [4]:
CLINICAL_REASONS = {'Disease progression', 'Death'}

reason = model_df['discontinuation_reason'].fillna('').str.strip()
model_df['Y'] = ((reason != '') & (~reason.isin(CLINICAL_REASONS))).astype(int)

print(model_df.groupby(['A', 'Y']).size().rename('n_patients'))
print()
print(f"Non-clinical discontinuation rate, A=1: {model_df.loc[model_df['A']==1, 'Y'].mean():.3f}")
print(f"Non-clinical discontinuation rate, A=0: {model_df.loc[model_df['A']==0, 'Y'].mean():.3f}")

A  Y
0  0    1364
   1     442
1  0     575
   1     342
Name: n_patients, dtype: int64

Non-clinical discontinuation rate, A=1: 0.373
Non-clinical discontinuation rate, A=0: 0.245


## Final data prep

`zepid`'s models take a patsy-style formula referencing raw columns directly (`C(col)` expands a
categorical internally), so no manual one-hot encoding is needed here, unlike the Cox/RSF
notebooks. Same `bsa_m2`-missing handling as before (3 patients, count printed).

In [5]:
tmle_df = model_df[[
    'A', 'Y', 'age', 'bsa_m2', 'on_panitumumab',
    'sex', 'race_category', 'ecog_baseline', 'diagnosis_type', 'trial_id',
]].copy()

before = len(tmle_df)
tmle_df = tmle_df.dropna(subset=['bsa_m2']).reset_index(drop=True)
print(f"Dropped {before - len(tmle_df)} patients missing bsa_m2 ({len(tmle_df)} remaining)")

tmle_df['on_panitumumab'] = tmle_df['on_panitumumab'].astype(int)
tmle_df.head()

Dropped 3 patients missing bsa_m2 (2720 remaining)


,A,Y,age,bsa_m2,on_panitumumab,sex,race_category,ecog_baseline,diagnosis_type,trial_id
0,1,1,67.0,1.848561,1,Female,Black or African American,Fully active,Colon,NCT00115225
1,0,0,72.0,1.984296,0,Male,White or Caucasian,Symptoms but ambulatory,Colon,NCT00115225
2,1,1,78.0,2.136634,1,Male,White or Caucasian,Fully active,Colon,NCT00115225
3,1,0,66.0,1.530432,0,Female,Black or African American,Symptoms but ambulatory,Colon,NCT00115225
4,1,1,67.0,2.191573,1,Male,White or Caucasian,Fully active,Rectal,NCT00115225


## Fit TMLE

`exposure_model` is the propensity score model (predicts `A` from confounders); `outcome_model`
predicts `Y` from `A` plus the same confounders. TMLE's doubly-robust update step combines both,
so the final estimate stays consistent even if one of the two is misspecified — the property
Design Spec §5.3 calls out `causalml`/`zEpid` for specifically.

**Library compatibility bug, worked around below:** `zepid` (last updated for an older
pandas/numpy) tries to bound predicted probabilities with an in-place assignment
(`v[v < bounds] = bounds`). Since pandas 3.0's Copy-on-Write, `Series.values` returns a
**read-only** array by default — confirmed directly (`arr.flags.writeable == False`) — so that
in-place assignment raises `ValueError: assignment destination is read-only`. Not a modeling
issue; patched with a small wrapper that copies the array before `zepid`'s own bounding function
runs, rather than downgrading pandas/numpy and risking compatibility issues with `lifelines`/
`scikit-survival` elsewhere in this project.

**Second bug, while writing the patch itself:** `zepid.causal.doublyrobust.TMLE` refers to the
`TMLE` **class** (re-exported into the package's `__init__.py`), not the `TMLE.py` submodule —
Python's normal package-attribute lookup finds the class first because of the name collision
between the module file and the class defined inside it. `import zepid.causal.doublyrobust.TMLE as
tmle_module` silently bound `tmle_module` to the class, not the module, so
`tmle_module.probability_bounds` raised `AttributeError`. Fixed with `importlib.import_module`,
which returns the actual submodule object directly, bypassing the shadowing.

In [6]:
import importlib

tmle_module = importlib.import_module('zepid.causal.doublyrobust.TMLE')

_original_probability_bounds = tmle_module.probability_bounds

def _writable_probability_bounds(v, bounds):
    return _original_probability_bounds(np.array(v, copy=True), bounds)

tmle_module.probability_bounds = _writable_probability_bounds

from zepid.causal.doublyrobust import TMLE

confounders = "age + bsa_m2 + on_panitumumab + C(sex) + C(race_category) + C(ecog_baseline) + C(diagnosis_type) + C(trial_id)"

tmle = TMLE(tmle_df, exposure='A', outcome='Y')
tmle.exposure_model(confounders, print_results=False)
tmle.outcome_model(f"A + {confounders}", print_results=False)
tmle.fit()
tmle.summary()

                Targeted Maximum Likelihood Estimator                 
Treatment:        A               No. Observations:     2720                
Outcome:          Y               No. Missing Outcome:  0                   
g-Model:          Logistic        Missing Model:        None                
Q-Model:          Logistic       
Risk Difference:     -0.007
95.0% two-sided CI: (-0.029 , 0.015)
----------------------------------------------------------------------
Risk Ratio:          0.976
95.0% two-sided CI: (0.903 , 1.054)
----------------------------------------------------------------------
Odds Ratio:          0.966
95.0% two-sided CI: (0.867 , 1.077)


## Findings

**Raw (unadjusted) association looked meaningful:** patients with a serious AE had a non-clinical
discontinuation rate of 37.3%, vs. 24.5% for patients without one — a ~13 percentage point gap
that would look like a strong candidate for a retention intervention if taken at face value.

**Doubly-robust adjusted estimate: no significant effect.**
```
Risk Difference:  -0.007   95% CI: (-0.029, 0.015)
Risk Ratio:        0.976   95% CI: ( 0.903, 1.054)
Odds Ratio:        0.966   95% CI: ( 0.867, 1.077)
```
Once confounding is accounted for (age, sex, race, ECOG baseline, diagnosis type, treatment arm,
`trial_id`), the effect collapses to essentially zero, and every confidence interval straddles
the null (0 for risk difference, 1 for risk ratio/odds ratio). This is a genuine, defensible
negative result, not a modeling failure — it's exactly the outcome TMLE is designed to catch:
confounding (sicker patients at baseline are both more likely to have an SAE *and* more likely to
drop out for other reasons, independent of the SAE itself) can manufacture a raw association that
isn't causal. **This is the actual value of running the causal-inference stage at all** — a naive
predictive read of the RSF's feature importance (`n_ae_events`/`max_severity_grade` ranked as the
top drivers) could have been misread as "reducing AE burden will reduce dropout," and TMLE shows
that conclusion isn't supported once confounding is properly adjusted for.

**Implication for the agent layer (Design Spec §6):** per this analysis, "serious AE occurrence"
should **not** be flagged to coordinators as a confirmed causal driver of non-clinical dropout —
the honest brief for a flagged patient would need to say the association isn't statistically
distinguishable from confounding, not present it as an actionable causal finding. This is a live
example of the "explicit flag when the leading driver is non-modifiable [or, here, not causally
established]" logic Design Spec §6 calls for.

**Caveats on this specific estimate, stated openly:**
- **Sample size for a 13-point observed gap that isn't significant** suggests real uncertainty,
  not overwhelming evidence of no effect — this is "not detected," not "proven zero."
- **Parametric nuisance models** (logistic regression via `statsmodels`, not machine-learning-based
  nuisance estimation) may not capture the true confounding relationship if it's meaningfully
  non-linear — a risk RSF's non-linear results (§5.2) suggest is plausible here.
- **`trial_id`-as-covariate**, not genuine cross-fitting/cluster-robust variance (stated in the
  estimand section above) — a real simplification, not full rigor.
- **AE-timing precision**: `start_day` establishes SAE-before-dropout ordering at the day level,
  but doesn't rule out short-term reverse pathways (e.g. a patient already deciding to withdraw
  experiencing a documented SAE in the days immediately before formal discontinuation is recorded).

**This completes the Design Spec §5 three-stage analytics progression**: Cox PH (baseline,
C-index 0.555) → RSF (adds non-linear/AE-burden effects, held-out C-index 0.609) → TMLE (causal
question: is the RSF's top predictive driver actually causal? No, not once confounding is
adjusted for). Next per the spec: the agent layer (§6), which needs to reflect this TMLE result
honestly rather than treating RSF's predictive feature importance as causal justification.